Update 05 boye summary file.R

bpowers24 opened the pull request:
I updated the script per issue #29 

I exported out the SPS summary data file (it's in the Share Drive). I spot checked it and it looked okay on my end. 

Next Steps before accepting pull request: 
- [ ] Review code
- [x] Spot check the SPS file for accuracy and formatting

bpowers24 commented:
Because we want to preserve the other script version too, let's rename this one `05_Boye_Summary_File_with_outlier_flag.R` (We will call the other one `05_Boye_Summary_File_with_outlier_file.R`).

brieanne-forbes commented:
Issues
-  lookup (beginning of outlier deviation string) could match to anything within a column name (ie N_OUTLIER) could remove TN 
- This issue applies within files with the current set up 
- Tried to determine if we could use beginning of string with underscores to match but the lookup isnt always the beginning of the column name (RATE_OUTLIER and Respiration_Rate_mg_DO_per_L_per_H)
- 
Potential solution
- write column name in deviation database

bpowers24 commented:
@brieanne-forbes: this PR is ready for you to review again. See comments at the top of the summary script for how the summarizing look up works (it's everything we had talked about when we brainstormed). I rearranged a lot of things from the last time you reviewed, so you should probably look at all of it again. 

Also note the tests that I ran in `test-05_Boye_Summary_File_outlier_flag.R`. The biggest thing to note is the code starting at LINE 229. This shows you existing outliers that we have already used and what column names we're expecting them to map to. I also listed out all the other columns we have used that don't yet have an associated outlier flag. There's nothing you need to run or take action on in this script, but pointing it out to show you the tested capacities of the summary script. The test file is also the reason why I made the main actions in the script into functions. Although the functions aren't super necessary in the context of the summary script itself, it made it a lot easier for me to validate that the look-ups were matching. 

Also, the only script you need to review is `05_Boye_Summary_File_outlier_flag.R`. The other files in this PR were accidently committed to this branch and can be ignored.

brieanne-forbes commented:
- [ ] compare SPS result to published file 
- [ ] combine had 940 obs but combine remove outlier had 943 obs, seems weird? But setdiff returns nothing?
- [ ] **CRITICAL**: somehow at line 221 in the function, you are ending up with samples duplicated. Some even repeated 6 times. I found this by checking the SPS_0021-3 was being removed from the mean. It had a row where the value was NA but also had a row with the value. I think this might be from when you pivot longer based on methods deviation and then you end up with the value in both rows but only remove it where the outlier flag is. Should reevaluate to make sure mean missing reps is working correctly after this is fixed.
- [ ] line 340: should tell the user what to do with the printed df

brieanne-forbes reviewed (changes requested):
See other comment

bpowers24 commented:
@brieanne-forbes: I addressed your comments.
- I compared to SPS. The new script includes 2 new columns (stdev_B6CA_del13C_per_mil, stdev_B5CA_del13C_per_mil) that weren't included in v3. Was this a conscious decision we made? 
- I believe the difference in row counts related to the critical issue you found. After fixing the issue, I confirmed that the row numbers match between `combine` and `combine_remove_outlier`. 
- Good catch! You were correct, the data were duplicating incorrectly during the pivot longer. I fixed this by not pivoting the methods deviations longer. I split them out into multiple (wide) columns and did rowwise calculations. I checked my work against tests that checked that the number of rows in the output matches the number or rows in the input. And also that the samples (both the exact sample and the count of samples) in the input matches the output. 
- I added a print statement and commented out the code that displays the table with the renamed column headers.

brieanne-forbes reviewed (changes requested):
No critical issues, the output seems good, especially with your extra checks. Few small things to address and then I think we are good! Thanks for your work on this.

- [x] Yes stdev columns were purposely left out. Maybe there can be a user input when there is "stdev" in the column name, ask if it should be included in the summary?
- [ ] I got an error when running the code (line 275) because there was an existing summary file. Maybe you can add an error and break in the function to tell the user to remove the file?
- [ ] I wonder if we should move the functions to a separate script now since the script is so long. Obviously not critical so up to you on this.

bpowers24 commented:
@brieanne-forbes:
- I added a new section (clean up summary) to ask the user if they'd like to remove any columns. I tried to make it dynamic in case we want to remove other cols in the future, but listed cols we already know we commonly omit. If you'd like me to hard code the cols we already know about, I can do that instead. I'd recommend running that code to see if it fits what you'd want.
- I added a warning to tell the user to remove the file if it detects a summary file. 
- I think I'm going to leave the functions in the script. I don't think they'd be useful outside of the context of this script, so I'd rather not complicate it, although I agree it's very long lol.

brieanne-forbes commented on Data_Package_Documentation/Boye_RF_Formatting/05_Boye_Summary_File_outlier_flag.R:
```diff
@@ -330,6 +373,53 @@ summary <- summary_calculated %>%
+    # DO_Concentration_At_Incubation_Time_Zero
+
+# tell the user common columns we usually omit
+cat("We often remove the following columns from data package summary files: ")
+summary %>% 
+  select(contains("stdev"),
+         contains("Respiration_R_Squared"), 
+         contains("Respiration_R_Squared_Adj"),
+         contains("Respiration_p_value"),
+         contains("Total_Incubation_Time_Min"),
+         contains("Number_Points_In_Respiration_Regression"),
+         contains("Number_Points_Removed_Respiration_Regression"),
+         contains("DO_Concentration_At_Incubation_Time_Zero")) %>% 
+  colnames()
+
```
```suggestion
cat("We often remove the following columns from data package summary files: ",
summary %>% 
  select(contains("stdev"),
         contains("Respiration_R_Squared"), 
         contains("Respiration_R_Squared_Adj"),
         contains("Respiration_p_value"),
         contains("Total_Incubation_Time_Min"),
         contains("Number_Points_In_Respiration_Regression"),
         contains("Number_Points_Removed_Respiration_Regression"),
         contains("DO_Concentration_At_Incubation_Time_Zero")) %>% 
  colnames())

```

brieanne-forbes replied:
Suggest making it one call instead of two which separates them in the console

brieanne-forbes commented on Data_Package_Documentation/Boye_RF_Formatting/05_Boye_Summary_File_outlier_flag.R:
```diff
@@ -330,6 +373,53 @@ summary <- summary_calculated %>%
+  colnames()
+
+# provide list of columns to the user
+cat("Listing all columns in the summary file: ")
+cat(paste0(seq_along(colnames(summary)), ". ", colnames(summary)), sep = "\n")
```
```suggestion
cat("Listing all columns in the summary file: ",
cat(paste0(seq_along(colnames(summary)), ". ", colnames(summary)), sep = "\n"))
```

brieanne-forbes reviewed (changes requested):
- [ ] It seems more useful to make a data frame of the column names in the summary and then have a column that flags if we usually remove it. Then do view(df)
- [ ] Maybe ask first if you would like to remove columns indicated and then have a second one asking if there are other columns? 

If you do my suggestion above, the inline edits are irrelevant

bpowers24 commented:
Additional notes on Brie's request: 

Rather than showing a list of commonly removed, then a list of all columns, and asking which numbers to remove, change to view a data frame. The data frame will have 3 columns: index, column_name, and default_remove (T/F). 

Ask the user: "do you want to remove the columns already selected? (Y/N)". If yes, then remove all of those columns. 

Then ask the user: "Which (additional) columns do you want to remove? Write 0 if none". View same df again. 

Using that df, remove the indicated columns.

bpowers24 commented:
@brieanne-forbes: I made the requested changes. If there are cols we usually drop, I display a df with the cols listed and ask the user if they would like to remove the ones already selected. Then I ask if they would like to drop any additional cols (if no default drop cols were present, then it skips to that second prompt). 
![image](https://github.com/user-attachments/assets/8ad7a5e5-4916-442a-a3a2-c0ee80d49b7a)

I update the df according to user prompts and then drop cols based on index.
